bdns/problem.go

```diff
@@ -15,6 +15,90 @@ type Error struct {
 	// Exactly one of rCode or underlying should be set.
 	underlying error
 	rCode      int
+
+	// Optional: If the resolver returned extended error information, it will be stored here.
+	// https://www.rfc-editor.org/rfc/rfc8914
+	extended *dns.EDNS0_EDE
+}
+
+// extendedDNSError returns non-nil if the input message contained an OPT RR
+// with an EDE option. https://www.rfc-editor.org/rfc/rfc8914.
+func extendedDNSError(msg *dns.Msg) *dns.EDNS0_EDE {
+	for _, extra := range msg.Extra {
+		if extra.Header().Rrtype != dns.TypeOPT {
+			continue
+		}
+
+		opt, ok := extra.(*dns.OPT)
+		if !ok {
+			continue
+		}
```

Comment:
```suggestion
	opt := msg.IsEdns0()
	if opt != nil {
```

Turns out there's a function to handle this bit of logic already

Reply:
Actually, that's maybe not exactly equivalent:  Can we have multiple OPT records?  if so this suggestion would only check if the first one is an EDE

Reply:
https://datatracker.ietf.org/doc/html/rfc6891#section-6.1.1

> When an OPT RR is included within any DNS message, it MUST be the only OPT RR in that message.

Reply:
Nice find, thanks!